### PR TITLE
EuroToken I: Tokkie (EuroToken's Tikkie)

### PR DESCRIPTION
**Pull request by Nieuwlaar**
EuroToken Group I has made a Tikkie-alike functionality where the user of the superapp can request a Tokkie. The requester can request money in either Euros or EuroTokens to someone, the requested can then pay with both Euros and EuroTokens (excluding Euros to Euros transactions, you can use Tikkie for that). The swapping between Euros and EuroTokens is done through the EuroToken Gateway.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,22 +22,36 @@ PeerChat implements a fully functional prototype of a distributed messaging app.
...
+You can send and receive digital Euros using QR-codes or build-in chat. **Experimental**.
+Sending Euros is as easy as sending a smiley.
+We did a test with native implementation of Trustchain and [a digital Euro last week](https://twitter.com/TriblerTeam/status/1367526077422256128).
+Field test date: 4 March 2021 at 10:30am.
+The native Android implementation in Kotlin is slowly getting mature.
+Location: the bar Doerak (with a liquor license! This is a special place, therefore selected as the site for our trail.
+Shops which sell coffee or closed canisters of alcohol are "essential shops" and therefore open in Corona times.) Loading real money on your phone requires an operational an open source [gateway](https://github.com/rwblokzijl/stablecoin-exchange) of Euros to digital Euros.
+Discussed in this master thesis issue: https://github.com/Tribler/tribler/issues/4629
+
+#### Double Spending mitigation
```
Comment (devos50) [inline README.md:40]: This does not seem to be part of your contribution?

File: app/src/README-eurotoken-payment-request.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+- Link Signature validation
+
+
+We had difficulties with setting E2E intergration tests since you will need to setup an instance of IPv8. We have also tried to do mock testing, however we came across a variety of problems. 
+
+
+
+## Support:
+If you come across issues with this feature or simply have implementation questions, you can send them to:
+t.l.nguyen@student.tudeflt.nl
```
Comment (devos50) [inline app/src/README-eurotoken-payment-request.md:106]: Your e-mail address wil be invalidated after you graduate. I would leave out this section :)

File: app/src/main/res/xml/network_security_config.xml [hunk 3]
```diff
@@ -3,7 +3,10 @@
     <domain-config cleartextTrafficPermitted="true">
         <!--        This is for bitcoin faucet, to claim some starter money using a REST call-->
         <domain includeSubdomains="true">134.122.59.107</domain>
+        <domain includeSubdomains="true">192.168.0.106</domain>
+        <domain includeSubdomains="true">80.112.144.229</domain>
```
Comment (devos50) [inline app/src/main/res/xml/network_security_config.xml:7]: Where do these IPs point to?

File: valuetransfer/src/androidTest/java/nl/tudelft/trustchain/valuetransfer/ui/exchangelink/ExchangeTransferMoneyLinkFragmentTest.kt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package nl.tudelft.trustchain.valuetransfer.ui.exchangelink
+
+// class ExchangeTransferMoneyLinkFragmentTest {
```
Comment (devos50) [inline valuetransfer/src/androidTest/java/nl/tudelft/trustchain/valuetransfer/ui/exchangelink/ExchangeTransferMoneyLinkFragmentTest.kt:3]: Is this file used? If not, you can just remove it I guess?

File: valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ValueTransferMainActivity.kt [hunk 5]
```diff
@@ -92,11 +91,12 @@ class ValueTransferMainActivity : BaseActivity() {
      * All fragments within this application, contact chat fragment excluded because it depends on arguments
      */
     private val fragmentManager = supportFragmentManager
-    private val walletOverviewFragment = WalletOverviewFragment()
+    private val walletOverviewFragment = ExchangeFragment()
```
Comment (devos50) [inline valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ValueTransferMainActivity.kt:94]: Is this change intended? If so, why?

File: valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ValueTransferMainActivity.kt [hunk 6]
```diff
@@ -192,14 +195,26 @@ class ValueTransferMainActivity : BaseActivity() {
...
-            .add(R.id.container, identityFragment, identityFragmentTag).hide(identityFragment)
-            .add(R.id.container, exchangeFragment, exchangeFragmentTag).hide(exchangeFragment)
-            .add(R.id.container, contactsFragment, contactsFragmentTag).hide(contactsFragment)
-            .add(R.id.container, qrScanController, qrScanControllerTag).hide(qrScanController)
-            .add(R.id.container, settingsFragment, settingsFragmentTag).hide(settingsFragment)
-            .add(R.id.container, walletOverviewFragment, walletOverviewFragmentTag)
-            .commit()
+        if (!requestMoney)
+            fragmentManager.beginTransaction()
+                .add(R.id.container, identityFragment, identityFragmentTag).hide(identityFragment)
```
Comment (devos50) [inline valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ValueTransferMainActivity.kt:200]: I think you can extract the common elements out of this if-else construct, like:

```
manager = fragmentManager.beginTransaction().hide(...).hide(...)
if(!requestMoney) { manager.hide(...) }
else { manager.hide(...) }
```

File: valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ui/exchange/ExchangeFragment.kt.orig [hunk 7]
```diff
@@ -0,0 +1,358 @@
+package nl.tudelft.trustchain.valuetransfer.ui.exchange
```
Comment (devos50) [inline valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ui/exchange/ExchangeFragment.kt.orig:1]: What is the purpose of this file? It even seems to contains merge conflicts...

File: valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/util/SecurityUtil.kt [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        return try {
+            val split = publicKey.toString().split("||")
+            val spec = RSAPublicKeySpec(BigInteger(split[0]), BigInteger(split[1]))
+            KeyFactory.getInstance(KeyProperties.KEY_ALGORITHM_RSA).generatePublic(spec)
+        } catch (ex: Exception) {
+            null
+        }
+    }
+
+    fun sign(input: String, privateKey: PrivateKey): String {
```
Comment (devos50) [inline valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/util/SecurityUtil.kt:58]: Would it be possible to use the underlying signature primitives provided by IPv8? If not, why not?

File: valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/dialogs/ExchangeTransferMoneyLinkDialog.kt [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun showLink(host: String, paymentId: String) {
+        val link = getLink(host, paymentId)
+        val sendIntent: Intent = Intent().apply {
+            action = Intent.ACTION_SEND
+            putExtra(
+                Intent.EXTRA_TEXT,
+                "Would you please pay me €$transactionAmountText  via\n$link"
```
Comment (devos50) [inline valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/dialogs/ExchangeTransferMoneyLinkDialog.kt:226]: Can you extract these strings to a resource file?

**Review by devos50 — CHANGES_REQUESTED** [review]
Thanks for your PR! I left a few comments. Please extract strings to a dedicated `strings.xml` file as much as possible. Also, review your commit history to pinpoint where unintended changes have been included.

Finally, please cleanup your commit history 👍

File: valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ui/exchange/ExchangeFragment.kt.orig (hunk 7 above)
Comment (InvictusRMC) [reply valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ui/exchange/ExchangeFragment.kt.orig:1]: > `.orig`

Appears to be some sort of backup.

**Comment by devos50** [comment]
Your branch seems to conflict with the `kotlin-ipv8` submodule. Don't forget to fix this when improving your PR 👍

File: valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/util/SecurityUtil.kt (hunk 8 above)
Comment (Nieuwlaar) [reply valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/util/SecurityUtil.kt:58]: Yes, added it to suggestions in the readme.

File: valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ui/exchange/ExchangeFragment.kt.orig (hunk 7 above)
Comment (Nieuwlaar) [reply valuetransfer/src/main/java/nl/tudelft/trustchain/valuetransfer/ui/exchange/ExchangeFragment.kt.orig:1]: Removed.

**Comment by Nieuwlaar** [comment]
>

File: eurotoken/src/main/java/nl/tudelft/trustchain/eurotoken/community/EuroTokenCommunity.kt [hunk 10]
```diff
@@ -41,7 +41,6 @@ class EuroTokenCommunity(
         transactionRepository.attemptRollback(peer, payload.transactionHash)
     }
 
-    @RequiresApi(Build.VERSION_CODES.O)
```
Comment (InvictusRMC) [inline eurotoken/src/main/java/nl/tudelft/trustchain/eurotoken/community/EuroTokenCommunity.kt:44]: Does this code now work on older devices?

Comment (InvictusRMC) [reply eurotoken/src/main/java/nl/tudelft/trustchain/eurotoken/community/EuroTokenCommunity.kt:91]: Hi team can you answer this question so we can hopefully merge your code?

**Comment by Nieuwlaar** [comment]
No, this was removed by accident. Thanks for the heads up. It was reverted in the latest commit.

File: app/src/README-eurotoken-payment-request.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+
+## Notes
+
+To make the process of link sharing and receiving the transactions easier, we have made some changes to the exchange, namely, we added a method called ```check_payments_done``` in the ```backend/stablecoin/stablecoin.py```. This method checks every 5 seconds for pending transactions to see if they are paid. If so, it will process the payment to finalize the transactions. This is especially relevant when the user uses the second scenario, namely Euro to EuroToken in which a Tikkie link will be used to pay in Euro.
+
+However, this change is not necessary when the server is deployed. When the server is deployed a webhook should be implemented. The webhook will be called by the bank (e.g., in case of Tikkie link, the bank is ABN AMRO) when the user has finished the payment and paid, and then the exchange server can process the transaction.
+
+
+## Notes
```
Comment (devos50) [inline app/src/README-eurotoken-payment-request.md:154]: Not sure what happened here but this seems to be an exact duplicate of the paragraph above.

File: eurotoken/src/main/java/nl/tudelft/trustchain/eurotoken/community/EuroTokenCommunity.kt [hunk 12]
```diff
@@ -87,7 +87,7 @@ class EuroTokenCommunity(
     private fun onRollbackRequest(peer: Peer, payload: RollbackRequestPayload) {
         transactionRepository.attemptRollback(peer, payload.transactionHash)
     }
-
+
```
Comment (devos50) [inline eurotoken/src/main/java/nl/tudelft/trustchain/eurotoken/community/EuroTokenCommunity.kt:90]: Minor comment: please avoid/remove these small cosmetic changes. They can potentially lead to rebase conflicts with other PRs. I also noticed a few added newlines in the main `README.md` that were not needed.

**Review by devos50 — CHANGES_REQUESTED** [review]

**Review by devos50 — APPROVED** [review]
Great, thank you 👍
